rllab/dynamics_randomization/variation.py

```diff
@@ -0,0 +1,143 @@
+from enum import Enum
+
+
+class VariationMethods(Enum):
+    COEFFICIENT = 1
+    ABSOLUTE = 2
+
+
+class VariationDistributions(Enum):
+    GAUSSIAN = 1
+    UNIFORM = 2
+
+
+class Variation:
+    def __init__(self):
+        self._xpath = None
+        self._attrib = None
+        self._method = None
+        self._distribution = None
+        self._var_range = None
+        self._elem = None
+        self._default = None
+
+    @property
+    def xpath(self):
+        return self._xpath
+
+    @xpath.setter
+    def xpath(self, xpath):
+        self._xpath = xpath
+
+    @property
+    def elem(self):
+        return self._elem
+
+    @elem.setter
+    def elem(self, elem):
+        self._elem = elem
+
+    @property
+    def attrib(self):
+        return self._attrib
+
+    @attrib.setter
+    def attrib(self, attrib):
+        self._attrib = attrib
+
+    @property
+    def default(self):
+        return self._default
+
+    @default.setter
+    def default(self, default):
+        self._default = default
+
+    @property
+    def method(self):
+        return self._method
+
+    @method.setter
+    def method(self, method):
+        self._method = method
```

Comment:
Variation objects should be immutable (i.e. there are no setters)

Reply:
Sorry, I just noticed this. There's a trade-off between having the fluent interface pattern and having immutable properties.
In order to set the members of the Variation object without setters, we have to pass their attributes in the constructor, and that breaks the interface pattern. Otherwise, without the setter decorators and with the members that correspond to each property not previously defined, we get this kind of errors:
```
Traceback (most recent call last):
  File "rllab/dynamics_randomization/trpo_swimmer.py", line 13, in <module>
    at_xpath(".//geom[@name='torso']").\
  File "/home/aigonzal/ivanWorkspace/rllab/rllab/dynamics_randomization/variation.py", line 114, in at_xpath
    self._list[-1].xpath = xpath
AttributeError: can't set attribute
```
Another way to do it is by setting the members of the class Variation directly in Variations, but that breaks the immutability of the property.
Any advice of how to achieve this is well appreciated.

Reply:
I was trying to use a dictionary to store the variation then add it to Variation class at last. But I found that we need the `v.elem=e` in MujocoModelGenerator, this needs to call the `elem` setter in Variation. So I didn’t think of a way to implement immutable Variation after all.

Reply:
For how to implement it, see [builder pattern](https://en.wikipedia.org/wiki/Builder_pattern#Python).

To solve the problem with `v.elem = e`, the solution is to keep a local cache of those elems in MujocoModelGenerator, rather than storing the data inside variations. That would be the proper way to do it--I just used the dict in my example as a shortcut.

Note that you can use the variation object directly as a key in your dictionary, i.e.
```python
elem_cache = {}

# caching elems
for v in variations:
    # find elem, = e
    elem_cache[v] = e

# making models
for v in variations:
    e = elem_cache[v]
```

Reply:
Thanks for the comment. Done